scripts/tutorial-uploader/lib/api.ts

```diff
@@ -0,0 +1,260 @@
+//
+// Any modifications or derivative works of this code must retain this
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+import { readFile } from "fs/promises";
+import { tmpdir } from "os";
+import { dirname, basename } from "path";
+import { randomBytes } from "crypto";
+
+import { $ } from "zx";
+import {
+  type RestClient,
+  createDirectus,
+  rest,
+  staticToken,
+  readItems,
+  readItem,
+  updateItem,
+  createItem,
+  deleteItem,
+  uploadFiles,
+} from "@directus/sdk";
+
+import { type LearningApiSchema, StringKeyOf, ElementType } from "./schema";
+import { type LocalTutorialData } from "./local-tutorial-data";
+
+/* To do:
+ *   [x] Fix types
+ *   [ ] Throw correctly on request failures
+ *   [ ] More helpful console logging
+ */
+
+export class API {
+  client: RestClient<LearningApiSchema>;
+
+  constructor(url: string, token: string) {
+    this.client = createDirectus<LearningApiSchema>(url)
+      .with(rest())
+      .with(staticToken(token));
+  }
+
+  /**
+   * Get IDs of all items in `collection` that match a field value.
+   * Roughly: "SELECT * FROM collection WHERE field=value".
+   */
+  async getIds<
+    CollectionName extends StringKeyOf<LearningApiSchema>,
+    FieldName extends StringKeyOf<
+      ElementType<LearningApiSchema[CollectionName]>
+    >,
+    FieldValue extends ElementType<
+      LearningApiSchema[CollectionName]
+    >[FieldName],
+  >(
+    collection: CollectionName,
+    field: FieldName,
+    value: FieldValue,
+  ): Promise<string[]> {
+    const response = await this.client.request(
+      // @ts-ignore
+      readItems(collection, { fields: ["id", field] }),
+    );
+    const matchingIds = response
+      .filter((item: any) => item[field] === value)
+      .map((item: any) => item.id);
+    return matchingIds;
+  }
+
+  /**
+   * Like getIds (plural) but expects at most one match.
+   * Throws if more than one match is found.
+   */
+  async getId<
+    CollectionName extends StringKeyOf<LearningApiSchema>,
+    FieldName extends StringKeyOf<
+      ElementType<LearningApiSchema[CollectionName]>
+    >,
+    FieldValue extends ElementType<
+      LearningApiSchema[CollectionName]
+    >[FieldName],
+  >(
+    collection: CollectionName,
+    field: FieldName,
+    value: FieldValue,
+  ): Promise<string | null> {
+    const ids = await this.getIds(collection, field, value);
+    if (ids.length === 0) {
+      return null;
+    }
+    if (ids.length === 1) {
+      return ids[0];
+    }
+    throw new Error(
+      `Found ${ids.length} items for getId('${collection}', '${field}', '${value}'. ` +
+        `Expected one or none.`,
+    );
+  }
+
+  /**
+   * Tutorials can have many translations, but we only use English at the moment.
+   */
+  async getEnglishTranslationId(tutorialId: string): Promise<number> {
+    // TODO: This assumes the only translation is english (currently true)
```

Comment:
No need to handle multiple yet, but please do throw an exception so we know we need to deal with it

Reply:
Fixed in 1696a8be75493290fafb945706a43689f6bfbc20; it now ignores non-English translations.